commtest/src/main/scala/coop/rchain/kademlia/kademlia.scala

```diff
@@ -0,0 +1,135 @@
+package coop.rchain.kademlia
+
+import scala.collection.mutable
+
+trait Keyed {
+  def key: Array[Byte]
+}
+
+trait Remote {
+  def ping: Unit
+}
+
+trait Peer extends Remote with Keyed
+
+trait Latent {
+  def latency: Double
+}
+
+trait Reputable {
+  def reputation: Double
+}
+
+object LatencyOrder extends Ordering[Latent] {
+  def compare(a: Latent, b: Latent) = a.latency compare b.latency
+}
+
+object ReputationOrder extends Ordering[Reputable] {
+  def compare(a: Reputable, b: Reputable) = a.reputation compare b.reputation
+}
+
+class PeerTableEntry[A <: Keyed](val entry: A) extends Keyed {
+  override def key = entry.key
+
+  override def toString = s"#{PeerTableEntry $entry}"
+}
+
+object PeerTable {
+  // Number of bits considered in the distance function. Taken from the
+  // passed-in "home" value to the table.
+  //
+  // val width = 256
+
+  // Maximum length of each row of the routing table.
+  val redundancy = 20
+
+  // Concurrency factor: system allows up to alpha outstanding network
+  // requests at a time.
+  val alpha = 3
+
+  // UNIMPLEMENTED: this parameter controls an optimization that can
+  // reduce the number hops required to find an address in the network
+  // by grouping keys in buckets of a size larger than one.
+  //
+  // val bucketWidth = 1
+}
+
+class PeerTable[A <: Peer](home: A,
+                           val k: Int = PeerTable.redundancy,
+                           val alpha: Int = PeerTable.alpha) {
+
+  type Entry = PeerTableEntry[A]
+
+  val width = home.key.size // in bytes
+  val table = Array.fill(8 * width) {
+    new mutable.MutableList[Entry]
+  }
+  // val byLatency = mutable.PriorityQueue.empty(LatencyOrder.reverse)
+  // val byReputation = mutable.PriorityQueue.empty(ReputationOrder)
+  val pending = new mutable.HashMap[Array[Byte], (A, A)]
+
+  private def ping(older: A, newer: A): Unit = {
+    // TODO constrain to alpha in flight
+    println(s"STUB: Pinging $older, might replace with $newer.")
+    pending synchronized {
+      pending.get(older.key) match {
+        case Some(_) => {
+          // Protocol error -- need throttling
+          println(s"Multiple pings in flight for $older.")
+        }
+        case None => {
+          pending(older.key) = (older, newer)
+          older.ping
+        }
+      }
+    }
+  }
+
+  // Kademlia XOR distance function.
+  def distance(a: A, b: A): Option[Int] = {
```

Comment:
You should try to avoid `return` in here.

Reply:
Ha. Originally it had no returns (except the legitimately non-local exit buried deep in the loops), and then we got a document full of Scala idioms, and it argues in favor of returns for guard conditions.

Reply:
I don't agree with the document on that one.

While here it is kind of okay, I think in general we should discourage the usage of `return`.
Martin Odersky thinks the same: https://twitter.com/propensive/status/838788224923418624

For a bunch of other reasons see: https://tpolecat.github.io/2014/05/09/return.html

Reply:
@pyrocto Do you agree on that we should remove the advice on using `return`?

Reply:
The blog post explains that 'return' exits from the enclosing method, not merely the enclosing function.  It makes it clear that 'return' is a very restricted kind of 'throw', so I would support its use in, for example, a library where the API requires error codes rather than exceptions. I don't think it's appropriate elsewhere: either use exceptions, option to indicate failure, or case classes to indicate what kind of failure.